solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,403 @@
+                    .getEntity().getContent());
+
+            ObjectReader reader = new ObjectMapper().reader(HashMap.class);
+            Map<String, Object> outerMap = reader.readValue(inputStream);
+
+            Map<String, Object> schemaMap = null;
+
+            if (outerMap != null) {
+                schemaMap = (Map) outerMap.get("schema");
+            }
+
+            if (schemaMap != null) {
+                List<Map<String, String>> fieldMapList = (List) schemaMap
+                        .get("fields");
+                List<String> columnNames = new ArrayList<String>();
+                List<ColumnType> columnTypes = new ArrayList<ColumnType>();
+
+                for (int i = 0; i < fieldMapList.size(); i++) {
+                    Map<String, String> fieldMap = fieldMapList.get(i);
+
+                    columnNames.add(fieldMap.get("name"));
+                    columnTypes.add(ColumnType.STRING);
+                }
+
+                String[] columnNamesArr = new String[columnNames.size()];
+                columnNamesArr = columnNames.toArray(columnNamesArr);
+
+                ColumnType[] columnTypesArr = new ColumnType[columnTypes.size()];
+                columnTypesArr = columnTypes.toArray(columnTypesArr);
+
+                tableDef = new SimpleTableDef(indexName, columnNamesArr,
+                        columnTypesArr);
+            }
+        } catch (IOException e) {
+            logger.error("Failed to parse schema", "", e);
+        }
+
+        return tableDef;
+    }
+
+    @Override
+    protected Schema getMainSchema() throws MetaModelException {
+        try {
+            final MutableSchema theSchema = new MutableSchema(
+                    getMainSchemaName());
+
+            final MutableTable table = tableDef.toTable().setSchema(theSchema);
+            final Column idColumn = table.getColumnByName(FIELD_ID);
+            if (idColumn != null && idColumn instanceof MutableColumn) {
+                final MutableColumn mutableColumn = (MutableColumn) idColumn;
+                mutableColumn.setPrimaryKey(true);
+            }
+            theSchema.addTable(table);
+
+            return theSchema;
+        } catch (Exception e) {
+            throw new MetaModelException("Schema retrieval failed " + e);
+        }
+    }
+
+    @Override
+    protected Number executeCountQuery(Table table,
+            List<FilterItem> whereItems, boolean functionApproximationAllowed) {
+        String queryStr = "*:*";
+
+        if (whereItems != null) {
+            for (int i = 0; i < whereItems.size(); i++) {
+                FilterItem whereItem = whereItems.get(i);
+                queryStr = whereItem.toString();
+            }
+        }
+
+        QueryResponse response = selectRows(table, null, queryStr, 0, 0);
+        return response.getResults().getNumFound();
+    }
+
+    @Override
+    public DataSet executeQuery(Query q) throws MetaModelException {
```

Comment:
The SQL interface works well if you do not override this method at all. The method is already implemented in the superclass and will delegate to some of the protected methods that you can choose to override. It is also possible to override this method, but it comes at a much greater risk becase the Query object is so multi-facetted that your code really has to take a ton of scenarios into account. I'll try and make line-notes below so you can see...